lib/Futures/Promise.h

```diff
@@ -1,163 +0,0 @@
-///
-/// Unless required by applicable law or agreed to in writing, software
-/// distributed under the License is distributed on an "AS IS" BASIS,
-/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-/// See the License for the specific language governing permissions and
-/// limitations under the License.
-///
-/// Copyright holder is ArangoDB GmbH, Cologne, Germany
-///
-/// @author Simon Grätzer
-////////////////////////////////////////////////////////////////////////////////
-
-#pragma once
-
-#include "Futures/Exceptions.h"
-#include "Futures/SharedState.h"
-#include "Futures/Unit.h"
-
-namespace arangodb {
-namespace futures {
-
-template<typename T>
-class Future;
-
-/// producer side of future-promise pair
-/// accesses on Promise have to be synchronized externally to
-/// be thread-safe
-template<typename T>
-class Promise {
-  static_assert(!std::is_same<void, T>::value, "Promise<Unit> instead of void");
-
- public:
-  /// make invalid promise
-  static Promise<T> makeEmpty() { return Promise(nullptr); }
-
-  /// @brief Constructs a Promise with no shared state.
-  /// After construction, valid() == true.
-  Promise() : _state(detail::SharedState<T>::make()), _retrieved(false) {}
-
-  Promise(Promise const& o) = delete;
-  Promise(Promise<T>&& o) noexcept
-      : _state(std::move(o._state)), _retrieved(o._retrieved) {
-    o._state = nullptr;
-  }
-
-  ~Promise() { this->detach(); }
-
-  Promise& operator=(Promise const&) = delete;
-  Promise& operator=(Promise<T>&& o) noexcept {
-    if (this != &o) {
-      detach();
-      _state = std::move(o._state);
-      _retrieved = o._retrieved;
-      o._retrieved = false;
-      o._state = nullptr;
-    }
-    return *this;
-  }
-
-  bool valid() const noexcept { return _state != nullptr; }
-
-  bool isFulfilled() const noexcept {
-    if (_state) {
-      return _state->hasResult();
-    }
-    return true;
-  }
-
-  /// Fulfill the Promise with an exception_ptr.
-  void setException(std::exception_ptr ep) { setTry(Try<T>(ep)); }
-
-  /// Fulfill the Promise with exception `e` *as if* by
-  ///   `setException(std::make_exception_ptr<E>(e))`.
-  template<class E>
-  typename std::enable_if<std::is_base_of<std::exception, E>::value>::type
-  setException(E const& e) {
-    setException(std::make_exception_ptr<E>(e));
-  }
-
-  /// Fulfill the Promise with the specified value using perfect forwarding.
-  /// Functionally equivalent to `setTry(Try<T>(std::forward<M>(value)))`
-  template<class M>
-  void setValue(M&& value) {
-    static_assert(!std::is_same<T, void>::value, "Use setValue() instead");
-    setTry(Try<T>(std::forward<M>(value)));
-  }
-
-  /// set void value
-  template<class B = T>
-  typename std::enable_if<std::is_same<Unit, B>::value>::type setValue() {
-    setTry(Try<Unit>(std::in_place));
-  }
-
-  /// Fulfill the Promise with the specified Try (value or exception).
-  void setTry(Try<T>&& t) {
-    throwIfFulfilled();
```

Comment:
I think the futures are written with the assumption that there will be exactly one user of the Promise, and exactly one user of the Future, so calling `set` from multiple threads would be an API violation.
So a thread holding the Promise may assume no other thread changes "has result", and a user of the Future may assume no other thread changes the callback.
So I'm not arguing against what you're saying, you're right, just that the implementation never tried to give that safety guarantee, and the "throwIfFulfilled" is just an opportunistic check against that (and one could argue it should be an assertion instead).

Reply:
> I think the futures are written with the assumption that there will be exactly one user of the Promise, and exactly one user of the Future, so calling set from multiple threads would be an API violation.

I agree

> So a thread holding the Promise may assume no other thread changes "has result", and a user of the Future may assume no other thread changes the callback.
So I'm not arguing against what you're saying, you're right, just that the implementation never tried to give that safety guarantee, and the "throwIfFulfilled" is just an opportunistic check against that (and one could argue it should be an assertion instead).

My point was about `hasResult()/isFulfilled` can used for sync, in that case it was incorrect, but because bad implementation we can skip it in tests.
Other words probability of data race if incorrect usage of api was lower
Also user can see atomic load in implementation and thought it is safe.
Another point after setTry/etc we don't need to hold shared state in promise (so we just make destruction&deallocation more lazy).